src/base/http/connection.h

```diff
@@ -27,41 +28,69 @@
  * exception statement from your version.
  */
 
-
 #pragma once
 
 #include <QElapsedTimer>
 #include <QObject>
 
+#include "types.h"
+
 class QTcpSocket;
 
 namespace Http
 {
-    class IRequestHandler;
-    struct Response;
-
     class Connection : public QObject
     {
         Q_OBJECT
         Q_DISABLE_COPY(Connection)
 
     public:
-        Connection(QTcpSocket *socket, IRequestHandler *requestHandler, QObject *parent = nullptr);
-        ~Connection();
+        enum Status
+        {
+            PendingRequest,
+            PendingStatus,
+            PendingHeaders,
+            PendingContent,
+            WritingContent,
+            Closed
+        };
+
+        // Connection takes ownership of socket
+        explicit Connection(QTcpSocket *socket, QObject *parent = nullptr);
 
-        bool hasExpired(qint64 timeout) const;
-        bool isClosed() const;
+        void sendStatus(const ResponseStatus &status);
+        void sendHeaders(const HeaderMap &headers);
+        void sendContent(const QByteArray &data);
+        void close();
+
+        Http::Environment enviroment() const;
+        Status status() const;
+        qint64 inactivityTime() const;
+        qint64 bytesToWrite() const;
+
+    signals:
+        void bytesWritten(qint64 bytes);
+        void disconnected();
+        void readyRequest(Request request);
 
     private slots:
         void read();
 
     private:
-        static bool acceptsGzipEncoding(QString codings);
-        void sendResponse(const Response &response) const;
+        enum ConnectionType
+        {
+            Unknown,
+            KeepAlive,
+            ClosedConnection
+        };
 
         QTcpSocket *m_socket;
-        IRequestHandler *m_requestHandler;
-        QByteArray m_receivedData;
         QElapsedTimer m_idleTimer;
+        QByteArray m_receivedData;
```

Comment:
Since now we are refactoring, please reserve memory for it (2~3 times of the max payload size).